File: __cpp__/algorithms/mergeSort.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> aux_vec;

void Merge(vector<int>& vec, int low, int mid, int high) {
	// assigning the left and right value for easier comparison
	int left = low;
	int right = mid+1;
	int aux_index = low;
	
	// pick the smaller no. between the left and right part
	while (left <= mid && right <= high) {
		if (vec[left] <= vec[right]) {
			aux_vec[aux_index] = vec[left];
			aux_index++;
			left++;
		}
		else {
			aux_vec[aux_index] = vec[right];
			aux_index += 1;
			right += 1;
		}
	}

	if (left <= mid) {
		for (int i=left; i <= mid; i++){
			aux_vec[aux_index] = vec[i];
			aux_index++;
		} 
	else {
		for (int i=right; i <= high; i++) {
			aux_vec[aux_index] = vec[i];
			aux_index++;
		}
	}

	// copy the sorted vec to the main vec
	for (int i=low; i <= high; i++) {
		vec[i] = aux_vec[i]
	}
}

void MergeSort(vector<int>& vec, int low, int high) {
	int mid;
	if (low < high) {
		mid = (low + high)/2;
		MergeSort(vec, low, mid);
		MergeSort(vec, mid+1, high);
		Merge(vec, low, mid, high);
	}
}

int main() {
	vector<int> vec = {7, 3, 5, 2, 4, 1, 8, 6, 0, 10, 9};

	// allocate space for the auxiliray vec
	aux_vec.resize(vec.size());

	MergeSort(vec, 0, vec.size()-1);

	cout << "Sorted Numbers: ";
	for (auto& it: vec) {
		cout << it << " ";
	}

	return 0;
}

File: __python__/__algorithms__/sorting algorithms/merge_sort.py
# MERGE SORT
from typing import List

def mergeSort(aux_array:List[int], array:List[int], low:int, high:int):
    if (low < high):
        mid = int((low + high)/2)
        mergeSort(aux_array, array, low, mid)  # recursively splits left half
        mergeSort(aux_array, array, mid+1, high)  # recursively splits right half
        merge(aux_array, array, low, mid, high)  # merge the left and right sorted halves


def merge(aux_array:List[int], array:List[int], low:int, mid:int, high:int):
    # assigning the left and right values for easier comparison
    left = low 
    right = mid+1
    aux_index = low
    # Pick the smaller number between the left part and the right part
    while (left <= mid and right <= high):
        if (array[left] <= array[right]):
            aux_array[aux_index] = array[left]
            aux_index += 1
            left += 1
        else:
            aux_array[aux_index] = array[right]
            aux_index += 1
            right += 1
        # print(aux_array)

    if (left <= mid):
        for i in range(left, mid + 1):
            aux_array[aux_index] = array[i]
            aux_index += 1
    else:
        for i in range(right, high + 1):
            aux_array[aux_index] = array[i]
            aux_index += 1
        
    # copy the sorted aux_array to the main array
    for i in range(low, high+1):
        array[i] = aux_array[i]


def main():
    array = [7, 3, 5, 2, 4, 1, 8, 6, 0, 10, 9]
    print("Unsorted array:\n{}".format(array))
    
    # Allocate space for the auxiliary vector
    aux_array = [None] * len(array)
    
    mergeSort(aux_array=aux_array, array=array, low=0, high=len(array)-1)
    print("\nSorted array:\n{}".format(array))

if __name__ == "__main__":
    main()